File: cpp_bind/softbody.cpp
#include "softbody.h"
#include "BulletSoftBody/btSoftBody.h"
#include "BulletSoftBody/btSoftBodyHelpers.h"
#include "BulletDynamics/Dynamics/btRigidBody.h"

nkSoftBodyWorldInfoHandle nk_softbody_world_info_create()
{
    btSoftBodyWorldInfo* info = new btSoftBodyWorldInfo();
    return static_cast<nkSoftBodyWorldInfoHandle>(info);
}

void nk_softbody_world_info_destroy(nkSoftBodyWorldInfoHandle info)
{
    if (!info) return;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    delete world_info;
}

void nk_softbody_world_info_set_gravity(nkSoftBodyWorldInfoHandle info, nkReal x, nkReal y, nkReal z)
{
    if (!info) return;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    world_info->m_gravity.setValue(x, y, z);
}

void nk_softbody_world_info_set_air_density(nkSoftBodyWorldInfoHandle info, nkReal density)
{
    if (!info) return;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    world_info->air_density = density;
}

void nk_softbody_world_info_set_water_density(nkSoftBodyWorldInfoHandle info, nkReal density)
{
    if (!info) return;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    world_info->water_density = density;
}

void nk_softbody_world_info_set_water_offset(nkSoftBodyWorldInfoHandle info, nkReal offset)
{
    if (!info) return;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    world_info->water_offset = offset;
}

void nk_softbody_world_info_set_water_normal(nkSoftBodyWorldInfoHandle info, nkReal x, nkReal y, nkReal z)
{
    if (!info) return;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    world_info->water_normal.setValue(x, y, z);
}

nkSoftBodyHandle nk_softbody_create_rope(nkSoftBodyWorldInfoHandle info, 
    nkReal from_x, nkReal from_y, nkReal from_z,
    nkReal to_x, nkReal to_y, nkReal to_z,
    int res, int fixeds)
{
    if (!info) return nullptr;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    btSoftBody* softbody = btSoftBodyHelpers::CreateRope(
        *world_info,
        btVector3(from_x, from_y, from_z),
        btVector3(to_x, to_y, to_z),
        res, fixeds
    );
    return static_cast<nkSoftBodyHandle>(softbody);
}

nkSoftBodyHandle nk_softbody_create_patch(nkSoftBodyWorldInfoHandle info,
    nkReal corner00_x, nkReal corner00_y, nkReal corner00_z,
    nkReal corner10_x, nkReal corner10_y, nkReal corner10_z,
    nkReal corner01_x, nkReal corner01_y, nkReal corner01_z,
    nkReal corner11_x, nkReal corner11_y, nkReal corner11_z,
    int resx, int resy, int fixeds, int gendiags)
{
    if (!info) return nullptr;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    btSoftBody* softbody = btSoftBodyHelpers::CreatePatch(
        *world_info,
        btVector3(corner00_x, corner00_y, corner00_z),
        btVector3(corner10_x, corner10_y, corner10_z),
        btVector3(corner01_x, corner01_y, corner01_z),
        btVector3(corner11_x, corner11_y, corner11_z),
        resx, resy, fixeds, gendiags
    );
    return static_cast<nkSoftBodyHandle>(softbody);
}

nkSoftBodyHandle nk_softbody_create_ellipsoid(nkSoftBodyWorldInfoHandle info,
    nkReal center_x, nkReal center_y, nkReal center_z,
    nkReal radius_x, nkReal radius_y, nkReal radius_z,
    int res)
{
    if (!info) return nullptr;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    btSoftBody* softbody = btSoftBodyHelpers::CreateEllipsoid(
        *world_info,
        btVector3(center_x, center_y, center_z),
        btVector3(radius_x, radius_y, radius_z),
        res
    );
    return static_cast<nkSoftBodyHandle>(softbody);
}

nkSoftBodyHandle nk_softbody_create_from_trimesh(nkSoftBodyWorldInfoHandle info,
    const nkReal* vertices, const int* triangles, int ntriangles)
{
    if (!info || !vertices || !triangles) return nullptr;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    
    int nvertices = 0;
    for (int i = 0; i < ntriangles * 3; i++)
    {
        if (triangles[i] + 1 > nvertices)
            nvertices = triangles[i] + 1;
    }
    
    btSoftBody* softbody = btSoftBodyHelpers::CreateFromTriMesh(
        *world_info,
        vertices,
        triangles,
        ntriangles,
        false
    );
    return static_cast<nkSoftBodyHandle>(softbody);
}

nkSoftBodyHandle nk_softbody_create_from_convex_hull(nkSoftBodyWorldInfoHandle info,
    const nkReal* vertices, int nvertices)
{
    if (!info || !vertices) return nullptr;
    btSoftBodyWorldInfo* world_info = static_cast<btSoftBodyWorldInfo*>(info);
    
    btAlignedObjectArray<btVector3> points;
    for (int i = 0; i < nvertices; i++)
    {
        points.push_back(btVector3(vertices[i * 3], vertices[i * 3 + 1], vertices[i * 3 + 2]));
    }
    
    btSoftBody* softbody = btSoftBodyHelpers::CreateFromConvexHull(
        *world_info,
        &points[0],
        nvertices
    );
    return static_cast<nkSoftBodyHandle>(softbody);
}

void nk_softbody_destroy(nkSoftBodyHandle softbody)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    delete body;
}

int nk_softbody_get_num_nodes(nkSoftBodyHandle softbody)
{
    if (!softbody) return 0;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    return body->m_nodes.size();
}

void nk_softbody_get_node_position(nkSoftBodyHandle softbody, int index, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!softbody || !out_x || !out_y || !out_z) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    if (index < 0 || index >= body->m_nodes.size()) return;
    
    const btSoftBody::Node& node = body->m_nodes[index];
    *out_x = node.m_x.x();
    *out_y = node.m_x.y();
    *out_z = node.m_x.z();
}

void nk_softbody_set_node_position(nkSoftBodyHandle softbody, int index, nkReal x, nkReal y, nkReal z)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    if (index < 0 || index >= body->m_nodes.size()) return;
    
    btSoftBody::Node& node = body->m_nodes[index];
    node.m_x.setValue(x, y, z);
}

void nk_softbody_get_node_velocity(nkSoftBodyHandle softbody, int index, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!softbody || !out_x || !out_y || !out_z) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    if (index < 0 || index >= body->m_nodes.size()) return;
    
    const btSoftBody::Node& node = body->m_nodes[index];
    *out_x = node.m_v.x();
    *out_y = node.m_v.y();
    *out_z = node.m_v.z();
}

void nk_softbody_set_node_velocity(nkSoftBodyHandle softbody, int index, nkReal x, nkReal y, nkReal z)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    if (index < 0 || index >= body->m_nodes.size()) return;
    
    btSoftBody::Node& node = body->m_nodes[index];
    node.m_v.setValue(x, y, z);
}

nkReal nk_softbody_get_node_mass(nkSoftBodyHandle softbody, int index)
{
    if (!softbody) return 0.0f;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    if (index < 0 || index >= body->m_nodes.size()) return 0.0f;
    
    const btSoftBody::Node& node = body->m_nodes[index];
    return node.m_im == 0 ? 0.0f : 1.0f / node.m_im;
}

void nk_softbody_set_node_mass(nkSoftBodyHandle softbody, int index, nkReal mass)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    if (index < 0 || index >= body->m_nodes.size()) return;
    
    btSoftBody::Node& node = body->m_nodes[index];
    node.m_im = mass == 0 ? 0 : 1.0f / mass;
}

nkReal nk_softbody_get_total_mass(nkSoftBodyHandle softbody)
{
    if (!softbody) return 0.0f;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    return body->getTotalMass();
}

void nk_softbody_set_total_mass(nkSoftBodyHandle softbody, nkReal mass)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->setTotalMass(mass);
}

void nk_softbody_set_velocity(nkSoftBodyHandle softbody, nkReal x, nkReal y, nkReal z)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->setVelocity(btVector3(x, y, z));
}

void nk_softbody_add_velocity(nkSoftBodyHandle softbody, nkReal x, nkReal y, nkReal z)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->addVelocity(btVector3(x, y, z));
}

void nk_softbody_apply_force(nkSoftBodyHandle softbody, nkReal x, nkReal y, nkReal z)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->addForce(btVector3(x, y, z));
}

void nk_softbody_apply_impulse(nkSoftBodyHandle softbody, nkReal x, nkReal y, nkReal z)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->addForce(btVector3(x, y, z));
}

void nk_softbody_clear_forces(nkSoftBodyHandle softbody)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    for (int i = 0; i < body->m_nodes.size(); i++)
    {
        body->m_nodes[i].m_f = btVector3(0, 0, 0);
    }
}

void nk_softbody_append_anchor(nkSoftBodyHandle softbody, int node_index, nkRigidBodyHandle body, 
    nkReal local_x, nkReal local_y, nkReal local_z, int disable_collision)
{
    if (!softbody || !body) return;
    btSoftBody* soft = static_cast<btSoftBody*>(softbody);
    btRigidBody* rigid = static_cast<btRigidBody*>(body);
    if (node_index < 0 || node_index >= soft->m_nodes.size()) return;
    
    soft->appendAnchor(node_index, rigid, btVector3(local_x, local_y, local_z), disable_collision != 0);
}

void nk_softbody_remove_anchor(nkSoftBodyHandle softbody, int node_index)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    if (node_index < 0 || node_index >= body->m_nodes.size()) return;
    
    for (int i = body->m_anchors.size() - 1; i >= 0; i--)
    {
        if (body->m_anchors[i].m_node == &body->m_nodes[node_index])
        {
            body->m_anchors.removeAtIndex(i);
            break;
        }
    }
}

void nk_softbody_set_material_stiffness(nkSoftBodyHandle softbody, nkReal kLST, nkReal kAST, nkReal kVST)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->m_materials[0]->m_kLST = kLST;
    body->m_materials[0]->m_kAST = kAST;
    body->m_materials[0]->m_kVST = kVST;
}

void nk_softbody_set_wind_velocity(nkSoftBodyHandle softbody, nkReal x, nkReal y, nkReal z)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->setWindVelocity(btVector3(x, y, z));
}

void nk_softbody_get_wind_velocity(nkSoftBodyHandle softbody, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!softbody || !out_x || !out_y || !out_z) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    btVector3 wind = body->getWindVelocity();
    *out_x = wind.x();
    *out_y = wind.y();
    *out_z = wind.z();
}

void nk_softbody_set_config_damping(nkSoftBodyHandle softbody, nkReal damping)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->m_cfg.kDP = damping;
}

void nk_softbody_set_config_drag(nkSoftBodyHandle softbody, nkReal drag)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->m_cfg.kDG = drag;
}

void nk_softbody_set_config_lift(nkSoftBodyHandle softbody, nkReal lift)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->m_cfg.kLF = lift;
}

void nk_softbody_set_config_pressure(nkSoftBodyHandle softbody, nkReal pressure)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->m_cfg.kPR = pressure;
}

void nk_softbody_set_config_volume_conversation(nkSoftBodyHandle softbody, nkReal volume)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->m_cfg.kVC = volume;
}

void nk_softbody_set_config_time_scale(nkSoftBodyHandle softbody, nkReal scale)
{
    if (!softbody) return;
    btSoftBody* body = static_cast<btSoftBody*>(softbody);
    body->m_cfg.timescale = scale;
}

File: rust/src/core/softbody.rs
use std::ffi::c_void;
use std::ptr::NonNull;

use super::rigidbody::RigidBody;
use super::types::{Real, Vec3};
use crate::ffi;

pub struct SoftBodyWorldInfo {
    handle: NonNull<c_void>,
}

impl SoftBodyWorldInfo {
    pub fn new() -> Self {
        let handle = unsafe { ffi::nk_softbody_world_info_create() };
        Self {
            handle: NonNull::new(handle).expect("Failed to create soft body world info"),
        }
    }

    pub fn set_gravity(&self, gravity: Vec3) {
        unsafe {
            ffi::nk_softbody_world_info_set_gravity(self.handle.as_ptr(), gravity.x, gravity.y, gravity.z);
        }
    }

    pub fn set_air_density(&self, density: Real) {
        unsafe {
            ffi::nk_softbody_world_info_set_air_density(self.handle.as_ptr(), density);
        }
    }

    pub fn set_water_density(&self, density: Real) {
        unsafe {
            ffi::nk_softbody_world_info_set_water_density(self.handle.as_ptr(), density);
        }
    }

    pub fn set_water_offset(&self, offset: Real) {
        unsafe {
            ffi::nk_softbody_world_info_set_water_offset(self.handle.as_ptr(), offset);
        }
    }

    pub fn set_water_normal(&self, normal: Vec3) {
        unsafe {
            ffi::nk_softbody_world_info_set_water_normal(self.handle.as_ptr(), normal.x, normal.y, normal.z);
        }
    }

    pub fn handle(&self) -> *mut c_void {
        self.handle.as_ptr()
    }
}

impl Default for SoftBodyWorldInfo {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for SoftBodyWorldInfo {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_softbody_world_info_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for SoftBodyWorldInfo {}
unsafe impl Sync for SoftBodyWorldInfo {}

pub struct SoftBody {
    handle: NonNull<c_void>,
}

impl SoftBody {
    pub fn create_rope(world_info: &SoftBodyWorldInfo, from: Vec3, to: Vec3, resolution: i32, fixeds: i32) -> Self {
        let handle = unsafe {
            ffi::nk_softbody_create_rope(
                world_info.handle(),
                from.x, from.y, from.z,
                to.x, to.y, to.z,
                resolution, fixeds,
            )
        };
        Self {
            handle: NonNull::new(handle).expect("Failed to create rope soft body"),
        }
    }

    pub fn create_patch(world_info: &SoftBodyWorldInfo, corners: [Vec3; 4], resx: i32, resy: i32, fixeds: i32, gendiags: bool) -> Self {
        let handle = unsafe {
            ffi::nk_softbody_create_patch(
                world_info.handle(),
                corners[0].x, corners[0].y, corners[0].z,
                corners[1].x, corners[1].y, corners[1].z,
                corners[2].x, corners[2].y, corners[2].z,
                corners[3].x, corners[3].y, corners[3].z,
                resx, resy, fixeds, if gendiags { 1 } else { 0 },
            )
        };
        Self {
            handle: NonNull::new(handle).expect("Failed to create patch soft body"),
        }
    }

    pub fn create_ellipsoid(world_info: &SoftBodyWorldInfo, center: Vec3, radius: Vec3, resolution: i32) -> Self {
        let handle = unsafe {
            ffi::nk_softbody_create_ellipsoid(
                world_info.handle(),
                center.x, center.y, center.z,
                radius.x, radius.y, radius.z,
                resolution,
            )
        };
        Self {
            handle: NonNull::new(handle).expect("Failed to create ellipsoid soft body"),
        }
    }

    pub fn create_from_trimesh(world_info: &SoftBodyWorldInfo, vertices: &[Real], triangles: &[i32]) -> Self {
        let handle = unsafe {
            ffi::nk_softbody_create_from_trimesh(
                world_info.handle(),
                vertices.as_ptr(),
                triangles.as_ptr(),
                (triangles.len() / 3) as i32,
            )
        };
        Self {
            handle: NonNull::new(handle).expect("Failed to create trimesh soft body"),
        }
    }

    pub fn create_from_convex_hull(world_info: &SoftBodyWorldInfo, vertices: &[Real]) -> Self {
        let handle = unsafe {
            ffi::nk_softbody_create_from_convex_hull(
                world_info.handle(),
                vertices.as_ptr(),
                (vertices.len() / 3) as i32,
            )
        };
        Self {
            handle: NonNull::new(handle).expect("Failed to create convex hull soft body"),
        }
    }

    pub fn handle(&self) -> *mut c_void {
        self.handle.as_ptr()
    }

    pub fn num_nodes(&self) -> i32 {
        unsafe { ffi::nk_softbody_get_num_nodes(self.handle.as_ptr()) }
    }

    pub fn get_node_position(&self, index: i32) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_softbody_get_node_position(self.handle.as_ptr(), index, &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn set_node_position(&self, index: i32, position: Vec3) {
        unsafe {
            ffi::nk_softbody_set_node_position(self.handle.as_ptr(), index, position.x, position.y, position.z);
        }
    }

    pub fn get_node_velocity(&self, index: i32) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_softbody_get_node_velocity(self.handle.as_ptr(), index, &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn set_node_velocity(&self, index: i32, velocity: Vec3) {
        unsafe {
            ffi::nk_softbody_set_node_velocity(self.handle.as_ptr(), index, velocity.x, velocity.y, velocity.z);
        }
    }

    pub fn get_node_mass(&self, index: i32) -> Real {
        unsafe { ffi::nk_softbody_get_node_mass(self.handle.as_ptr(), index) }
    }

    pub fn set_node_mass(&self, index: i32, mass: Real) {
        unsafe {
            ffi::nk_softbody_set_node_mass(self.handle.as_ptr(), index, mass);
        }
    }

    pub fn total_mass(&self) -> Real {
        unsafe { ffi::nk_softbody_get_total_mass(self.handle.as_ptr()) }
    }

    pub fn set_total_mass(&self, mass: Real) {
        unsafe {
            ffi::nk_softbody_set_total_mass(self.handle.as_ptr(), mass);
        }
    }

    pub fn set_velocity(&self, velocity: Vec3) {
        unsafe {
            ffi::nk_softbody_set_velocity(self.handle.as_ptr(), velocity.x, velocity.y, velocity.z);
        }
    }

    pub fn add_velocity(&self, velocity: Vec3) {
        unsafe {
            ffi::nk_softbody_add_velocity(self.handle.as_ptr(), velocity.x, velocity.y, velocity.z);
        }
    }

    pub fn apply_force(&self, force: Vec3) {
        unsafe {
            ffi::nk_softbody_apply_force(self.handle.as_ptr(), force.x, force.y, force.z);
        }
    }

    pub fn apply_impulse(&self, impulse: Vec3) {
        unsafe {
            ffi::nk_softbody_apply_impulse(self.handle.as_ptr(), impulse.x, impulse.y, impulse.z);
        }
    }

    pub fn clear_forces(&self) {
        unsafe {
            ffi::nk_softbody_clear_forces(self.handle.as_ptr());
        }
    }

    pub fn append_anchor(&self, node_index: i32, body: &RigidBody, local: Vec3, disable_collision: bool) {
        unsafe {
            ffi::nk_softbody_append_anchor(
                self.handle.as_ptr(),
                node_index,
                body.handle(),
                local.x, local.y, local.z,
                if disable_collision { 1 } else { 0 },
            );
        }
    }

    pub fn remove_anchor(&self, node_index: i32) {
        unsafe {
            ffi::nk_softbody_remove_anchor(self.handle.as_ptr(), node_index);
        }
    }

    pub fn set_material_stiffness(&self, k_lst: Real, k_ast: Real, k_vst: Real) {
        unsafe {
            ffi::nk_softbody_set_material_stiffness(self.handle.as_ptr(), k_lst, k_ast, k_vst);
        }
    }

    pub fn set_wind_velocity(&self, velocity: Vec3) {
        unsafe {
            ffi::nk_softbody_set_wind_velocity(self.handle.as_ptr(), velocity.x, velocity.y, velocity.z);
        }
    }

    pub fn wind_velocity(&self) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_softbody_get_wind_velocity(self.handle.as_ptr(), &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn set_config_damping(&self, damping: Real) {
        unsafe {
            ffi::nk_softbody_set_config_damping(self.handle.as_ptr(), damping);
        }
    }

    pub fn set_config_drag(&self, drag: Real) {
        unsafe {
            ffi::nk_softbody_set_config_drag(self.handle.as_ptr(), drag);
        }
    }

    pub fn set_config_lift(&self, lift: Real) {
        unsafe {
            ffi::nk_softbody_set_config_lift(self.handle.as_ptr(), lift);
        }
    }

    pub fn set_config_pressure(&self, pressure: Real) {
        unsafe {
            ffi::nk_softbody_set_config_pressure(self.handle.as_ptr(), pressure);
        }
    }

    pub fn set_config_volume_conversation(&self, volume: Real) {
        unsafe {
            ffi::nk_softbody_set_config_volume_conversation(self.handle.as_ptr(), volume);
        }
    }

    pub fn set_config_time_scale(&self, scale: Real) {
        unsafe {
            ffi::nk_softbody_set_config_time_scale(self.handle.as_ptr(), scale);
        }
    }
}

impl Drop for SoftBody {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_softbody_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for SoftBody {}
unsafe impl Sync for SoftBody {}
